src/methods.jl

```diff
@@ -1,5 +1,88 @@
 export FDMReport, fdm, backward_fdm, forward_fdm, central_fdm
 
+forward_grid(p::Int) = 0:(p - 1)
+backward_grid(p::Int) = (1 - p):0
+function central_grid(p::Int)
+    if isodd(p)
+        return div(1 - p, 2):div(p - 1, 2)
+    else
+        return vcat(div(-p, 2):-1, 1:div(p, 2))
+    end
+end
+
+abstract type FDMethod end
+
+for D in (:Forward, :Backward, :Central)
+    gridf = Symbol(lowercase(String(D)), "_grid")
+    @eval begin
+        struct $D{G<:AbstractVector, C<:AbstractVector} <: FDMethod
+            p::Int
+            q::Int
+            grid::G
+            coefs::C
+        end
+        function $D(p::Integer, q::Integer)
+            q < p || throw(ArgumentError("order of the method must be strictly greater " *
+                                         "than that of the derivative"))
+            # Check whether the method can be computed. We require the factorial of the
+            # method order to be computable with regular `Int`s, but `factorial` will overflow
+            # after 20, so 20 is the largest we can allow.
+            p > 20 && throw(ArgumentError("order of the method is too large to be computed"))
+
+            grid = $gridf(p)
+
+            C = [g^i for i in 0:(p - 1), g in grid]
+            x = zeros(Int, p)
+            x[q + 1] = factorial(q)
+            coefs = C \ x
+
+            return $D{typeof(grid), typeof(coefs)}(p, q, grid, coefs)
+        end
+        (d::$D)(f, x; kwargs...) = fdm(d, f, x; kwargs...)
+    end
+end
+
+maxabs(x) = maximum(abs, x)
+
+function fdm(
+    m::M,
+    f,
+    x;
+    bound=maxabs(f(x)),
+    eps=Base.eps(bound),
```

Comment:
We also have to be careful here that `eps > 0`. Again the simple solution is to add a tiny number, e.g. `1e-20`.

Reply:
`Base.eps` will always be nonzero, so I think in this case it should be sufficient to simply add a check that `eps > 0` and throw an error otherwise.

Reply:
Ah, that is true. I'm not sure if `Base.eps(0.0) == 5.0e-324` is too small though. We should check that.

Reply:
Too small in what sense? It's small enough that e.g. `1 / eps(0.0)` is `Inf`, which I suppose is problematic.

Reply:
I'm worried about the below computation on `h\hat`. In general, working near the edges of the floating point ranges is dangerous. For example `Base.eps(0.0) / 10 == 0.0`, so stuff could go very wrong if `C2` is large. Adaptation could make this even more complicated.

Reply:
Heh, yeah:
```
julia> Central(5, 1)(sin, 1; adapt=1, eps=eps(0.0))
-Inf
```